Packet++/src/Asn1Codec.cpp

```diff
@@ -0,0 +1,569 @@
+			tagLen = 2;
+		}
+
+		Asn1Record* newRecord;
+
+		if (isConstructed)
+		{
+			if (tagClass == Asn1TagClass::Universal)
+			{
+				switch (static_cast<Asn1UniversalTagType>(tagType))
+				{
+					case Asn1UniversalTagType::Sequence:
+					{
+						newRecord = new Asn1SequenceRecord();
+						break;
+					}
+					case Asn1UniversalTagType::Set:
+					{
+						newRecord = new Asn1SetRecord();
+						break;
+					}
+					default:
+					{
+						newRecord = new Asn1ConstructedRecord();
+					}
+				}
+			}
+			else
+			{
+				newRecord = new Asn1ConstructedRecord();
+			}
+		}
+		else
+		{
+			if (tagClass == Asn1TagClass::Universal)
+			{
+				auto asn1UniversalTagType = static_cast<Asn1UniversalTagType>(tagType);
+				switch (asn1UniversalTagType)
+				{
+					case Asn1UniversalTagType::Integer:
+					{
+						newRecord = new Asn1IntegerRecord();
+						break;
+					}
+					case Asn1UniversalTagType::Enumerated:
+					{
+						newRecord = new Asn1EnumeratedRecord();
+						break;
+					}
+					case Asn1UniversalTagType::OctetString:
+					{
+						newRecord = new Asn1OctetStringRecord();
+						break;
+					}
+					case Asn1UniversalTagType::Boolean:
+					{
+						newRecord = new Asn1BooleanRecord();
+						break;
+					}
+					case Asn1UniversalTagType::Null:
+					{
+						newRecord = new Asn1NullRecord();
+						break;
+					}
+					default:
+					{
+						newRecord = new Asn1GenericRecord();
+					}
+				}
+			}
+			else
+			{
+				newRecord = new Asn1GenericRecord();
+			}
+		}
+
+		newRecord->m_TagClass = tagClass;
+		newRecord->m_IsConstructed = isConstructed;
+		newRecord->m_TagType = tagType;
+
+		return newRecord;
+	}
+
+	int Asn1Record::decodeLength(const uint8_t* data, size_t dataLen)
+	{
+		if (dataLen < 1)
+		{
+			throw std::invalid_argument("Cannot decode ASN.1 record length");
+		}
+
+		// Check 8th bit
+		auto lengthForm = data[0] & 0x80;
+
+		auto numberLengthBytes = 1;
+
+		// Check if the tag is using more than one byte
+		// 8th bit at 0 means the length only uses one byte
+		// 8th bit at 1 means the length uses more than one byte. The number of bytes is encoded in the other 7 bits
+		if (lengthForm != 0)
+		{
+			auto additionalLengthBytes = data[0] & 0x7F;
+			if (static_cast<int>(dataLen) < additionalLengthBytes + 1)
+			{
+				throw std::invalid_argument("Cannot decode ASN.1 record length");
+			}
+			for (auto index = additionalLengthBytes; index > 0; --index)
+			{
+				m_ValueLength += data[index] * static_cast<int>(std::pow(256, (additionalLengthBytes - index)));
+			}
+			numberLengthBytes += additionalLengthBytes;
+		}
+		else
+		{
+			m_ValueLength = data[0];
+		}
+
+		return numberLengthBytes;
+	}
+
+	void Asn1Record::decodeValueIfNeeded()
+	{
+		if (m_EncodedValue != nullptr)
+		{
+			decodeValue(m_EncodedValue, true);
+			m_EncodedValue = nullptr;
+		}
+	}
+
+	Asn1GenericRecord::Asn1GenericRecord(Asn1TagClass tagClass, bool isConstructed, uint8_t tagType, const uint8_t* value, size_t valueLen)
```

Comment:
any reason not to use the member initializer lists?

Reply:
All members are part of the base class `Asn1Record` so we can't initialize them in a member initializer list: https://stackoverflow.com/questions/7405740/how-can-i-initialize-base-class-member-variables-in-derived-class-constructor